storage: TestStoreRangeMergeSlowUnabandonedFollower_WithSplit failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/bbc646fc6de90b59c0253fd682667715959fb657

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stressrace TESTS=TestStoreRangeMergeSlowUnabandonedFollower_WithSplit PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=993747&tab=buildLog

```
I181030 05:31:23.053635 45478 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:43819
W181030 05:31:23.115136 45260 gossip/gossip.go:1496  [n3] no incoming or outgoing connections
I181030 05:31:23.115472 45589 gossip/client.go:129  [n3] started gossip client to 127.0.0.1:43819
I181030 05:31:23.115619 45260 gossip/gossip.go:393  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:40959" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I181030 05:31:23.120530 45260 storage/store_snapshot.go:636  [s1,r1/1:/M{in-ax}] sending preemptive snapshot e89247b4 at applied index 16
I181030 05:31:23.120978 45260 storage/store_snapshot.go:679  [s1,r1/1:/M{in-ax}] streamed snapshot to (n2,s2):?: kv pairs: 49, log entries: 6, rate-limit: 2.0 MiB/sec, 1ms
I181030 05:31:23.121304 45605 storage/replica_raftstorage.go:798  [s2,r1/?:{-}] applying preemptive snapshot at index 16 (id=e89247b4, encoded size=8344, 1 rocksdb batches, 6 log entries)
I181030 05:31:23.121904 45605 storage/replica_raftstorage.go:804  [s2,r1/?:/M{in-ax}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I181030 05:31:23.122241 45260 storage/replica_command.go:853  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2, gen=0]
I181030 05:31:23.123953 45260 storage/replica.go:3935  [s1,r1/1:/M{in-ax},txn=b53c4525] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181030 05:31:23.134169 45260 storage/store_snapshot.go:636  [s1,r1/1:/M{in-ax}] sending preemptive snapshot c5743a91 at applied index 18
I181030 05:31:23.134629 45260 storage/store_snapshot.go:679  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 52, log entries: 8, rate-limit: 2.0 MiB/sec, 9ms
I181030 05:31:23.135196 45607 storage/replica_raftstorage.go:798  [s3,r1/?:{-}] applying preemptive snapshot at index 18 (id=c5743a91, encoded size=9286, 1 rocksdb batches, 8 log entries)
I181030 05:31:23.136167 45607 storage/replica_raftstorage.go:804  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I181030 05:31:23.136798 45260 storage/replica_command.go:853  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I181030 05:31:23.146812 45260 storage/replica.go:3935  [s1,r1/1:/M{in-ax},txn=1f432ec1] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I181030 05:31:23.421064 45260 storage/replica_command.go:299  [s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r2]
I181030 05:31:23.427460 45260 storage/replica_command.go:440  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0] into this range
I181030 05:31:23.458120 45300 storage/store.go:2562  [s1,r1/1:{/Min-b},txn=d84bc307] removing replica r2/1
I181030 05:31:23.458543 45403 storage/store.go:2562  [s2,r1/2:{/Min-b}] removing replica r2/2
I181030 05:31:23.459156 45260 storage/replica_command.go:299  [s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r3]
I181030 05:31:23.519369 45260 storage/replica_command.go:853  [s1,r1/1:{/Min-b}] change replicas (REMOVE_REPLICA (n3,s3):3): read existing descriptor r1:{/Min-b} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=3]
I181030 05:31:23.535956 45260 storage/replica.go:3935  [s1,r1/1:{/Min-b},txn=0237a7d2] proposing REMOVE_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2] next=4
I181030 05:31:23.537501 45629 storage/store.go:2562  [replicaGC,s3,r1/3:{/Min-b}] removing replica r1/3
I181030 05:31:23.537782 45629 storage/replica.go:873  [replicaGC,s3,r1/3:{/Min-b}] removed 14 (9+5) keys in 0ms [clear=0ms commit=0ms]
W181030 05:31:23.538975 45594 storage/raft_transport.go:584  while processing outgoing Raft queue to node 3: rpc error: code = Unavailable desc = transport is closing:
--- FAIL: TestStoreRangeMergeSlowUnabandonedFollower_WithSplit (0.51s)
	client_merge_test.go:1780: r1 was not found on s3

```

SHA: https://github.com/cockroachdb/cockroach/commits/bbc646fc6de90b59c0253fd682667715959fb657

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stressrace TESTS=TestStoreRangeMergeSlowUnabandonedFollower_WithSplit PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=993992&tab=buildLog

```
I181030 07:37:50.960063 45362 gossip/gossip.go:393  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:35939" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I181030 07:37:50.961991 45711 gossip/client.go:129  [n3] started gossip client to 127.0.0.1:41833
I181030 07:37:51.034169 45362 storage/store_snapshot.go:636  [s1,r1/1:/M{in-ax}] sending preemptive snapshot ce511a3c at applied index 16
I181030 07:37:51.035907 45362 storage/store_snapshot.go:679  [s1,r1/1:/M{in-ax}] streamed snapshot to (n2,s2):?: kv pairs: 49, log entries: 6, rate-limit: 2.0 MiB/sec, 4ms
I181030 07:37:51.037147 45621 storage/replica_raftstorage.go:798  [s2,r1/?:{-}] applying preemptive snapshot at index 16 (id=ce511a3c, encoded size=8346, 1 rocksdb batches, 6 log entries)
I181030 07:37:51.041362 45621 storage/replica_raftstorage.go:804  [s2,r1/?:/M{in-ax}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=2ms commit=1ms]
I181030 07:37:51.044503 45362 storage/replica_command.go:853  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2, gen=0]
I181030 07:37:51.053714 45362 storage/replica.go:3935  [s1,r1/1:/M{in-ax},txn=b2734468] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I181030 07:37:51.068989 45362 storage/store_snapshot.go:636  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 388a6670 at applied index 18
I181030 07:37:51.072194 45362 storage/store_snapshot.go:679  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 52, log entries: 8, rate-limit: 2.0 MiB/sec, 7ms
I181030 07:37:51.074832 45744 storage/replica_raftstorage.go:798  [s3,r1/?:{-}] applying preemptive snapshot at index 18 (id=388a6670, encoded size=9290, 1 rocksdb batches, 8 log entries)
I181030 07:37:51.080197 45744 storage/replica_raftstorage.go:804  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 5ms [clear=0ms batch=0ms entries=3ms commit=0ms]
I181030 07:37:51.086084 45362 storage/replica_command.go:853  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I181030 07:37:51.116179 45362 storage/replica.go:3935  [s1,r1/1:/M{in-ax},txn=9e1e0f2b] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I181030 07:37:51.285818 45362 storage/replica_command.go:299  [s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r2]
I181030 07:37:51.360152 45362 storage/replica_command.go:440  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0] into this range
W181030 07:37:51.422405 45103 storage/intent_resolver.go:675  [s1] failed to push during intent resolution: failed to push "merge" id=6356269d key=/Local/Range/Min/RangeDescriptor rw=true pri=0.00027870 iso=SERIALIZABLE stat=PENDING epo=0 ts=0.000000123,350 orig=0.000000123,350 max=0.000000123,350 wto=false rop=false seq=1
I181030 07:37:51.467381 45476 storage/store.go:2562  [s1,r1/1:{/Min-b},txn=6356269d] removing replica r2/1
I181030 07:37:51.468607 45535 storage/store.go:2562  [s2,r1/2:{/Min-b}] removing replica r2/2
I181030 07:37:51.471871 45362 storage/replica_command.go:299  [s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r3]
I181030 07:37:51.586074 45362 storage/replica_command.go:853  [s1,r1/1:{/Min-b}] change replicas (REMOVE_REPLICA (n3,s3):3): read existing descriptor r1:{/Min-b} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=3]
I181030 07:37:51.599765 45362 storage/replica.go:3935  [s1,r1/1:{/Min-b},txn=5f9a898a] proposing REMOVE_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2] next=4
I181030 07:37:51.618797 45377 storage/store.go:2562  [replicaGC,s3,r1/3:{/Min-b}] removing replica r1/3
I181030 07:37:51.620468 45377 storage/replica.go:873  [replicaGC,s3,r1/3:{/Min-b}] removed 14 (9+5) keys in 1ms [clear=1ms commit=0ms]
--- FAIL: TestStoreRangeMergeSlowUnabandonedFollower_WithSplit (0.91s)
	client_merge_test.go:1780: r1 was not found on s3

```

--- a/pkg/storage/client_merge_test.go
+++ b/pkg/storage/client_merge_test.go
@@ -1771,17 +1771,8 @@ func TestStoreRangeMergeSlowUnabandonedFollower_WithSplit(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// Remove and GC the LHS replica on store2. This ensures that the RHS replica
-	// on store2 will never be subsumed, because the merge trigger will never be
-	// applied by the LHS.
+	// Remove the LHS replica from store2.
 	mtc.unreplicateRange(lhsDesc.RangeID, 2)
-	lhsRepl2 := store2.LookupReplica(lhsDesc.StartKey)
-	if err := store2.ManualReplicaGC(lhsRepl2); err != nil {
-		t.Fatal(err)
-	}
-	if _, err := store2.GetReplica(lhsDesc.RangeID); err == nil {
-		t.Fatal("lhs replica not destroyed on store2")
-	}
 
 	// Transfer the lease on the new RHS to store2 and wait for it to apply. This
 	// will force its replica to of the new RHS to become up to date, which

